### DISCUSSION TITLE
openwrt上怎么加速所有使用CF的站点

### DISCUSSION THREAD

**Discussion by jsdfhasuh** [XIU2/CloudflareSpeedTest#302] [讨论]
或者怎么判断那些站点是套了CF的

**Comment by XIU2** [comment]
我没有 OpenWrt 系统设备，所以无法给予你该系统相关帮助。
反正我在 Windows 上用的是 3Proxy #71 ，原理就是 3Proxy 创建了一个本地 HTTP 代理，浏览器让所有流量走这个本地代理，然后 3Proxy 解析域名得到 IP，根据 IP 判断是不是 Cloudflare CDN，是的话就将其重定向至指定 IP。

有些 DNS 程序也能实现类似的功能。

**Comment by jsdfhasuh** [comment]
> 我没有 OpenWrt 系统设备，所以无法给予你该系统相关帮助。 反正我在 Windows 上用的是 3Proxy #71 ，原理就是 3Proxy 创建了一个本地 HTTP 代理，浏览器让所有流量走这个本地代理，然后 3Proxy 解析域名得到 IP，根据 IP 判断是不是 Cloudflare CDN，是的话就将其重定向至指定 IP。

怎么根据IP判断Cloudflare CDN，是不是有一个IP列表的

**Comment by XIU2** [comment]
Cloudflare CDN 的 IP 列表：
https://www.cloudflare.com/zh-cn/ips/

**Comment by miniers** [comment]
我搞了套mosdns来当dns，如果 域名dns请求 返回的 ip 在CF的ip段，则自动替换dns响应 ip为优选ip，试了下我这里效果不算好，经常tls握手不成功，还不如dns默认返回的cf ip 连接性好了，你要无聊可以试试
docker run -d -e ENABLE_CFST=true -v $(pwd)/config:/config -p 53:53/udp miniers/mosdns

项目地址:https://github.com/miniers/docker-mosdns

**Comment by jsdfhasuh** [comment]
我想了用iptables来实现，但是不行，在openwrt上写了规则，但nas没起作用，奇怪

**Comment by tofuliang** [comment]
> 我想了用iptables来实现，但是不行，在openwrt上写了规则，但nas没起作用，奇怪

```bash
ipset -N cf hash:net;
for i in $(</etc/config/ip.txt grep -oE "[0-9\.\/]+");do echo $i;ipset add cf $i;done;
iptables -t nat -I PREROUTING  ! -d 172.64.68.79 -m set --match-set cf dst -j DNAT --to-destination 172.64.68.79
```
/etc/config/ip.txt 就是cloudflare的IP段.
如果网络支持ipv6的话,需要禁用ipv6结果的返回(没写ipv6的规则),我是使用smartdns针对常用的域名禁止了返回ipv6.
感觉是生效了,但是看iptables统计出来的包数量和流量都很少.如果有不对的,望指教.

**Comment by nie11kun** [comment]
我是在 openwrt 上修改 hosts 使用的。

**Comment by 120318** [comment]
> > 我想了用iptables来实现，但是不行，在openwrt上写了规则，但nas没起作用，奇怪
> 
> ```shell
> ipset -N cf hash:net;
> for i in $(</etc/config/ip.txt grep -oE "[0-9\.\/]+");do echo $i;ipset add cf $i;done;
> iptables -t nat -I PREROUTING  ! -d 172.64.68.79 -m set --match-set cf dst -j DNAT --to-destination 172.64.68.79
> ```
> 
> /etc/config/ip.txt 就是cloudflare的IP段. 如果网络支持ipv6的话,需要禁用ipv6结果的返回(没写ipv6的规则),我是使用smartdns针对常用的域名禁止了返回ipv6. 感觉是生效了,但是看iptables统计出来的包数量和流量都很少.如果有不对的,望指教.

这个有什么最新的进展么~

**Comment by tofuliang** [comment]
> > > 我想了用iptables来实现，但是不行，在openwrt上写了规则，但nas没起作用，奇怪
> > 
> > 
> > ```shell
> > ipset -N cf hash:net;
> > for i in $(</etc/config/ip.txt grep -oE "[0-9\.\/]+");do echo $i;ipset add cf $i;done;
> > iptables -t nat -I PREROUTING  ! -d 172.64.68.79 -m set --match-set cf dst -j DNAT --to-destination 172.64.68.79
> > ```
> > 
> > /etc/config/ip.txt 就是cloudflare的IP段. 如果网络支持ipv6的话,需要禁用ipv6结果的返回(没写ipv6的规则),我是使用smartdns针对常用的域名禁止了返回ipv6. 感觉是生效了,但是看iptables统计出来的包数量和流量都很少.如果有不对的,望指教.
> 
> 这个有什么最新的进展么~

在Openwrt下使用`tcpdump -i pppoe-wan ip host 172.64.68.79 and dst port 80`命令验证过了，可行。
在Openwrt下的设备执行 `telnet IP 80`,请求IP符合`cf`ipset的话，都会请求到`172.64.68.79`这个IP。

**Comment by ihuanx** [comment]
> > > > 我想了用iptables来实现，但是不行，在openwrt上写了规则，但nas没起作用，奇怪
> > > 
> > > 
> > > ```shell
> > > ipset -N cf hash:net;
> > > for i in $(</etc/config/ip.txt grep -oE "[0-9\.\/]+");do echo $i;ipset add cf $i;done;
> > > iptables -t nat -I PREROUTING  ! -d 172.64.68.79 -m set --match-set cf dst -j DNAT --to-destination 172.64.68.79
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > /etc/config/ip.txt 就是cloudflare的IP段. 如果网络支持ipv6的话,需要禁用ipv6结果的返回(没写ipv6的规则),我是使用smartdns针对常用的域名禁止了返回ipv6. 感觉是生效了,但是看iptables统计出来的包数量和流量都很少.如果有不对的,望指教.
> > 
> > 
> > 这个有什么最新的进展么~
> 
> 在Openwrt下使用`tcpdump -i pppoe-wan ip host 172.64.68.79 and dst port 80`命令验证过了，可行。 在Openwrt下的设备执行 `telnet IP 80`,请求IP符合`cf`ipset的话，都会请求到`172.64.68.79`这个IP。

再具体一点呗..  看不懂这两条命令 IP段丢哪里

**Comment by tofuliang** [comment]
> ```shell
> ipset -N cf hash:net;
> for i in $(</etc/config/ip.txt grep -oE "[0-9\.\/]+");do echo $i;ipset add cf $i;done;
> iptables -t nat -I PREROUTING  ! -d 172.64.68.79 -m set --match-set cf dst -j DNAT --to-destination 172.64.68.79
> ```
> 
> /etc/config/ip.txt 就是cloudflare的IP段. 如果网络支持ipv6的话,需要禁用ipv6结果的返回(没写ipv6的规则),我是使用smartdns针对常用的域名禁止了返回ipv6. 感觉是生效了,但是看iptables统计出来的包数量和流量都很少.如果有不对的,望指教.

@ihuanx 这个就是具体命令了，上面`tcpdump`那个只是验证是否成功的。

**Comment by ihuanx** [comment]
@tofuliang  `tcpdump` 测试似乎是转发成功了 不过我实际测试了 好像并没有生效 F12里看的IP也是默认分配的IP

**Comment by tofuliang** [comment]
> `tcpdump` 测试似乎是转发成功了 不过我实际测试了 好像并没有生效 F12里看的IP也是默认分配的IP

这个过程对于浏览器来说应该是透明的.可以进一步做下测试,直接把`默认分配的IP`直接Block掉,看看是否能正常访问.
